ludwig/utils/entmax/activations.py

```diff
@@ -190,7 +188,18 @@ def backward(cls, ctx, dY):
         return dX, None, None
 
 
-def sparsemax(X, dim=-1, k=None):
+def _entmax15_forward(X, dim, k):
+    max_val, _ = X.max(dim=dim, keepdim=True)
+    X = X - max_val  # same numerical stability trick as for softmax
+    X = X / 2  # divide by 2 to solve actual Entmax
+
+    tau_star, _ = _entmax_threshold_and_support(X, dim=dim, k=k)
+
+    Y = torch.clamp(X - tau_star, min=0) ** 2
+    return Y, {}
+
+
+def sparsemax(X, dim=-1, k=None, training=True):
```

Comment:
Couldn't find it here, but from where does this `training` param get plumbed through to this sparsemax call?

Reply:
This is coming from the `nn.Module` wrapper for each of the entmax modes: https://github.com/ludwig-ai/ludwig/blob/9a2efa6865a26bc1b7e95487d910c8723c06be6c/ludwig/utils/entmax/activations.py#L274

Reply:
Ah, then this must be set when we call `model.train()` vs `model.eval()`. Got it, thanks!